### PR TITLE
Universal reader output base classes

### PR DESCRIPTION
**Pull request by JoOkuma**
Hi, this is a work in progress for the base classes so we can advance with the universal reader.

I'm happy to iterate and get feedback from every to have a friendly interface.

I think we could be more restrictive than ome-zarr in some aspects; this would reduce the number of edge cases we need to worry about when implementing our own pipelines (e.g. fixing the shape to 5 and 6)

### CODE REVIEW

File: iohub/dataset_base.py
Comment (JoOkuma) [reply iohub/dataset_base.py]: I was a bit more strict. I added the types supported by zarr. What do you think?

**Comment by JoOkuma** [comment]
> @JoOkuma We don't need `__setitem__` and `__delitem__` since these are read-only.
> 
> Also just to clarify I was talking about `BaseFOVCollection`. So we can do `bool("name" in SomeFOVCollection())`.
> 
> For `BaseFOV` I think it might be ok to ignore pyramids for the time being. (In theory TIFFs can have pyramids too.)

Got it, I updated it accordingly.

File: iohub/dataset_base.py
Comment (ziw-liu) [reply iohub/dataset_base.py]: What doe ellipsis in `Union` mean?

File: iohub/dataset_base.py
Comment (ziw-liu) [inline iohub/dataset_base.py:56]:
```suggestion
        """Check if a position is present in the collection."""
```

File: iohub/dataset_base.py
Comment (JoOkuma) [reply iohub/dataset_base.py]: The ellipsis is for the tuple, the syntax is a bit misleading.
With python 3.11 it can be typed as, which is clearer in my opinion:

```python3
int | slice | tuple[int | slice, ...]
```

It can be an int, a slice, or a tuple with ints or slices.

Comment (ziw-liu) [reply iohub/dataset_base.py]: Oops wrong bracket count. Thanks for the explanation!

File: iohub/dataset_base.py
Comment (ziw-liu) [inline iohub/dataset_base.py:70]: `NGFFNode` also has `__iter__`, but that only iterates on the immediate next level in the hierarchy (e.g. rows in plate, wells in row etc). The `rows()`, `wells()`, and `positions()` methods allows for iteration on different levels from the same starting point (e.g. plate).

Comment (ziw-liu) [reply iohub/dataset_base.py:70]: I think it is still a good idea to always yield `BaseFOV` here. So we will need to change the behavior of the NGFF plate node for compliance.

File: iohub/dataset_base.py
Comment (ziw-liu) [inline iohub/dataset_base.py:44]: You can get `DTypeLike` from `numpy.typing`. So we don't need the np namespace in this file.

File: iohub/dataset_base.py
Comment (ziw-liu) [inline iohub/dataset_base.py:33]:
```suggestion
        Returned object must support the ``__array__`` interface,
        so that ``np.asarray(...)`` will work.
```
Sphinx formatting.

File: iohub/dataset_base.py
Comment (ziw-liu) [inline iohub/dataset_base.py:48]: Should there be a constraint on the length? It's a tricky thing since NGFF has scale in channel dimension, which has no meaning and should always be 1. The time 'scale' is a bit ambiguous too...

Comment (ziw-liu) [reply iohub/dataset_base.py:48]: We can also just do `spatial_scale` here and it will always be a 3-tuple.

File: iohub/dataset_base.py
Comment (ziw-liu) [inline iohub/dataset_base.py:53]:
```suggestion
class BaseFOVCollection(ABC):
    @abstractmethod
    def __enter__(self) -> BaseFOVCollection:
        "Open the underlying file and return self."
        raise NotImplementedError

    @abstractmethod
    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        "Close the files."
        raise NotImplementedError
```

Can we also have context manager here?

File: iohub/dataset_base.py
Comment (JoOkuma) [reply iohub/dataset_base.py:44]: `DTypeLike` is a bit broader than `np.dtype`, for example, the string `"unit8"` is `DTypeLike`, if that's OK with you I'm fine with switching.

File: iohub/dataset_base.py
Comment (JoOkuma) [reply iohub/dataset_base.py:80]: Ok, can I leave this to be implemented in a different PR when we inherit the base classes on the `NGFFNode`?

File: iohub/dataset_base.py
Comment (ziw-liu) [reply iohub/dataset_base.py:45]: IMO it's fine since that seems to be how most array libraries (numpy, zarr, torch etc) type their API. But stricter typing doesn't harm so I don't have a strong opinion here.

File: iohub/dataset_base.py
Comment (ziw-liu) [inline iohub/dataset_base.py:9]: Then what is the intended way to access voxel values?

File: iohub/dataset_base.py
Comment (JoOkuma) [reply iohub/dataset_base.py:49]: I'll do `zyx_scale`, it's more informative, IMO.

File: iohub/dataset_base.py
Comment (ziw-liu) [reply iohub/dataset_base.py:80]: Sounds good! Actually I should take care of it once this merges.

File: iohub/dataset_base.py
Comment (ziw-liu) [reply iohub/dataset_base.py:9]: p.s. the `data` property is not exactly 'intended' either...

Comment (JoOkuma) [reply iohub/dataset_base.py:9]: I thought of accessing through `__getitem__` and having additional methods to access the multi-scale in ome-zarr datasets. We would have to have a policy to select an array when multiple keys are present.

Comment (ziw-liu) [reply iohub/dataset_base.py:10]: In practice most of the acquisition data formats (TIFF, clear control) do not have multi-scale. Should the key always be '0' for them?

Comment (ziw-liu) [reply iohub/dataset_base.py:10]: Oh wait did you mean that `__getitem__` directly slices the lowest resolution array? So getting the whole FOV would be `BaseFOV(...)[:]`?

File: iohub/dataset_base.py
Comment (ziw-liu) [inline iohub/dataset_base.py:87]:
```suggestion
        """Iterates over pairs of keys and FOVs."""
```

File: iohub/dataset_base.py
Comment (JoOkuma) [reply iohub/dataset_base.py:10]: I would slice the **highest** resolution.

It would behave sort of like a zarr.
I'm not completely sure if it should load into memory or allow returning any kind of array (e.g. dask array).

I'm more in favor of the latter, and the user should call ```np.asarray``` when it must be loaded into memory, which is not always the case because `numpy` should process any `arraylike` data.

Comment (ziw-liu) [reply iohub/dataset_base.py:10]: Got it, so internally it would be like this (for NGFF)?

```py
class Position(...):
    def __getitem__(self, key):
        return self["0"][key]
```

Comment (JoOkuma) [reply iohub/dataset_base.py:10]: Yes, we could even allow any key, and use "0" or the first in lexicographic order when more than one key is present.

Comment (ziw-liu) [reply iohub/dataset_base.py:10]: I don't think we currently use any other format with pyramids or other data structure that needs multiple array keys in the same FOV. However it does call for more change in the NGFF API since it also does writing, since if `__getitem__` operates on one of the arrays in the FOV, `__setitem__` should match that behavior.

Comment (JoOkuma) [reply iohub/dataset_base.py:10]: Yes, it will need to be refactored.

File: iohub/dataset_base.py
Comment (ziw-liu) [inline iohub/dataset_base.py:25]: If we fix this to a 5-tuple, should `__getitem__` match this behavior? i.e. if an FOV has an underlying array of 3 dimensions, should the slicing be `[0,0,0]` or `[0,0,0,0,0]` to get a scalar?

**Comment by ziw-liu** [comment]
Also we should figure out how the consequent PRs should merge. It's fine if this one merges into main but if we keep the refactoring of readers atomic and distributed, asynchronously merging into main will result in a broken default branch.

Maybe make a new target branch (e.g. `uapi-readers`) once this is merged?

File: iohub/dataset_base.py
Comment (keithchev) [reply iohub/dataset_base.py:25]: From the perspective of non-expert end users, I think it would be confusing if the pixel values in a 3D image had to be accessed using something like `[x, y, z, 0, 0]`. (It would be even worse if the placeholder channel and time dimensions came first, because then a naive indexing using `[x, y, z]` would not only not result in a scalar but would be entirely incorrect)

**Comment by JoOkuma** [comment]
> Also we should figure out how the consequent PRs should merge. It's fine if this one merges into main but if we keep the refactoring of readers atomic and distributed, asynchronously merging into main will result in a broken default branch.
> 
> Maybe make a new target branch (e.g. `uapi-readers`) once this is merged?

I like the idea of having a separate branch.

I'm happy to take a few days next week and be done with it. I'm free after next Wednesday, but I'll need help with `tiff` because I'm not super familiar with it.

File: iohub/dataset_base.py
Comment (keithchev) [reply iohub/dataset_base.py:25]: Fair enough!

Comment (ziw-liu) [reply iohub/dataset_base.py:25]: There is a glitch in slicing with all 5 dimensions though. Translating the slicing makes it impossible to return a lazy Zarr array, removing many useful attributes (append, chunks, fancy slicing etc) that we currently inherit from `zarr.Array`. And at writing time, a NumPy array is directly written to Zarr, preserving shape, and it would be very confusing to have to slice them differently (or even having two different array types depending on how the file is opened).

Comment (JoOkuma) [reply iohub/dataset_base.py:25]: Uhm... an alternative is to implement a method on `BaseFOV` that translate the 5-dim indexing to the original array, and when requesting the whole array we can call `np.expand_dims`.

Comment (ziw-liu) [reply iohub/dataset_base.py:25]: > Uhm... an alternative is to implement a method on `BaseFOV` that translate the 5-dim indexing to the original array, and when requesting the whole array we can call `np.expand_dims`.

Yeah I think this is reasonable. It actually has to call `expand_dims` even when not requesting the whole array. So we might as well just delegate this to dask or tensorstore by constructing the reshaped always-lazy da/ts array with `dask()` or `tensorstore()` as we currently have with `ngff.ImageArray.numpy()`.

Comment (ziw-liu) [reply iohub/dataset_base.py:25]: > So we might as well just delegate this to dask or tensorstore by constructing the reshaped always-lazy da/ts array with `dask()` or `tensorstore()` as we currently have with `ngff.ImageArray.numpy()`.

This way we can also wrap dask in xarray and delegate the dimension naming API too.

**Review by edyoshikun — COMMENTED** [review]
This all LGTM! Thank you guys.  I like the `BaseCollectionFOV` as an iterable that we can use to access individual positions. 

A couple of points just to make sure this will behave as intended. 
- For this universal reader, we will call `BaseFOV[0,0,0]` this will  return a (1,1,1,Z,Y,X) 5D array?

**Comment by ziw-liu** [comment]
> * For this universal reader, we will call `BaseFOV[0,0,0]` this will  return a (1,1,1,Z,Y,X) 5D array?

My understanding is that it will behave as normal array slicing does. So ZYX (3D) in this case.

File: iohub/dataset_base.py
Comment (ziw-liu) [inline iohub/dataset_base.py:118]: This constructor needs reference to all of the FOVs at creation time, meaning that for HCS OME-Zarr the entire tree has to be walked upon reading the plate. Is there a way to do this lazily as it is currently?

File: iohub/dataset_base.py
Comment (ziw-liu) [reply iohub/dataset_base.py]: @JoOkuma and I had a chat and we will try out zarr's `view` for this.

File: iohub/dataset_base.py
Comment (JoOkuma) [reply iohub/dataset_base.py:118]: HCS OME-Zarr could have its own constructor and provide no parameter do the parent `FOVCollection` class, or do you prefer an abstract base class that doesn't initialize anything?

Comment (ziw-liu) [reply iohub/dataset_base.py:118]: But then the attributes initialized are expected elsewhere, which makes it difficult to maintain. I think the base class should be abstract and the implementation here can become something like `FOVDict`

File: iohub/dataset_base.py
Comment (talonchandler) [inline iohub/dataset_base.py:25]: Maybe `channels` -> `channel_names`, since `channels` might make you think you're getting the number of channels?

File: iohub/dataset_base.py
Comment (talonchandler) [inline iohub/dataset_base.py:93]: I like this choice.

File: iohub/dataset_base.py
Comment (talonchandler) [inline iohub/dataset_base.py:100]: `BaseFOV` has `zyx_scale` and `channels/channel_names`, but no property related to the time dimension. 

Maybe `def t_scale(self) -> float`? Or maybe we want to support irregular time steps with `def times(self) -> list[float]`?

File: iohub/dataset_base.py
Comment (talonchandler) [inline iohub/dataset_base.py:25]: Is there a "correct" way to enforce that `len(BaseFOV.channels)` should be identical to `BaseFOV.shape[1]` within the abstract base class?

Similarly, if we include a `BaseFOV.times` then `len(BaseFOV.times)` should be identical to `BaseFOV.shape[0]`.

**Review by talonchandler — APPROVED** [review]
Looks great to me! I'm excited. 

Thanks for pushing on this @JoOkuma.

File: iohub/dataset_base.py
Comment (JoOkuma) [reply iohub/dataset_base.py:100]: Do you this the `list[float]` should be enforced by the API, or is it very specific to some datasets?

File: iohub/dataset_base.py
Comment (JoOkuma) [reply iohub/dataset_base.py:25]: I'm not really sure what's the best practice. I think that should be covered by tests, another option is to add an assert in this class but it will only be useful if `super().channels` is called.

File: iohub/dataset_base.py
Comment (JoOkuma) [reply iohub/dataset_base.py:100]: @ziw-liu @talonchandler What do you think about ditching `zyx_scale` and having `scale` that returns a 5-tuple of floats?

**Comment by JoOkuma** [comment]
@ziw-liu, I renamed files from `database_base.py` to `fov.py`.

File: iohub/fov.py
Comment (ziw-liu) [inline iohub/fov.py:184]: Is this redundant if `__init__` already receives a `dict`?

File: iohub/dataset_base.py
Comment (JoOkuma) [reply iohub/dataset_base.py]: @ziw-liu , I see. I was expecting `Position.scale` to return the same, but it might not be 5-dimensional, right? This might lead to confusion in the future.

File: iohub/fov.py
Comment (JoOkuma) [reply iohub/fov.py:184]: It can be initialized in two different ways, so this additional check is necessary, and we only pay a very small overhead.

- kwargs init: `FOVDict(green=..., red=...)`
- dict init: `FOVDict({"green": ..., "red": ...})`
- both: `FOVDict({"green": ...}, red=...)`

File: iohub/fov.py
Comment (JoOkuma) [reply iohub/fov.py:222]: It should return `False`; thanks for noticing that.
[reference](https://stackoverflow.com/a/71171297/6748803)

File: iohub/fov.py
Comment (ziw-liu) [inline iohub/fov.py:74]:
```suggestion
        if len(seq) != len(_AXES_PREFIX):
            raise RuntimeError(f"Failed to pad raw axes {self.axes_names} to {_AXES_PREFIX}")
```

**Comment by JoOkuma** [comment]
Thanks y'all.

**Pull request merged**
